### ISSUE TITLE
conch client ignores --known-hosts option for writing (but honors it for reading)

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#3496]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)| @glyph reported|
|-|-|
|Trac ID|trac#3496|
|Type|enhancement|
|Created|2008-10-24 15:26:41Z|

A demonstrative traceback:

```
glyph@alastor:~/.ssh$ ls known_hosts
ls: cannot access known_hosts: No such file or directory
glyph@alastor:~/.ssh$ conch --known-hosts alt_hosts localhost echo hi
The authenticity of host '127.0.0.1' can't be established.
RSA key fingerprint is 1f:2c:e1:01:ed:77:65:fd:8b:fb:5e:b4:60:d8:dd:70.
Are you sure you want to continue connecting (yes/no)? yes
Warning: Permanently added '127.0.0.1' (RSA) to the list of known hosts.
Traceback (most recent call last):
  File "/home/glyph/Projects/Twisted/trunk/twisted/python/log.py", line 51, in callWithLogger
    return callWithContext({"system": lp}, func, *args, **kw)
  File "/home/glyph/Projects/Twisted/trunk/twisted/python/log.py", line 36, in callWithContext
    return context.call({ILogContext: newCtx}, func, *args, **kw)
  File "/home/glyph/Projects/Twisted/trunk/twisted/python/context.py", line 59, in callWithContext
    return self.currentContext().callWithContext(ctx, func, *args, **kw)
  File "/home/glyph/Projects/Twisted/trunk/twisted/python/context.py", line 37, in callWithContext
    return func(*args,**kw)
--- <exception caught here> ---
  File "/home/glyph/Projects/Twisted/trunk/twisted/internet/selectreactor.py", line 146, in _doReadOrWrite
    why = getattr(selectable, method)()
  File "/home/glyph/Projects/Twisted/trunk/twisted/internet/tcp.py", line 463, in doRead
    return self.protocol.dataReceived(data)
  File "/home/glyph/Projects/Twisted/trunk/twisted/conch/ssh/transport.py", line 314, in dataReceived
    self.dispatchMessage(messageNum, packet[1:])
  File "/home/glyph/Projects/Twisted/trunk/twisted/conch/ssh/transport.py", line 329, in dispatchMessage
    f(payload)
  File "/home/glyph/Projects/Twisted/trunk/twisted/conch/ssh/transport.py", line 996, in ssh_KEX_DH_GEX_REPLY
    d = self.verifyHostKey(pubKey, fingerprint)
  File "/home/glyph/Projects/Twisted/trunk/twisted/conch/client/direct.py", line 91, in verifyHostKey
    fingerprint)
  File "/home/glyph/Projects/Twisted/trunk/twisted/conch/client/default.py", line 46, in verifyHostKey
    known_hosts = open(os.path.expanduser('~/.ssh/known_hosts'), 'r+')
exceptions.IOError: [Errno 2] No such file or directory: '/home/glyph/.ssh/known_hosts'
```'

This will probably be fixed by my branch for [#10942](https://github.com/twisted/twisted/issues/10942).

<details><summary>Searchable metadata</summary>

```
trac-id__3496 3496
type__enhancement enhancement
reporter__glyph glyph
priority__normal normal
milestone__ 
branch__ 
branch_author__ 
status__closed closed
resolution__fixed fixed
component__conch conch
keywords__ 
time__1224862001000000 1224862001000000
changetime__1226804104000000 1226804104000000
version__None None
owner__ 

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_1"></a>|@glyph set status to `closed`|
|-|-|

(In [25365]) Fixes for the conch client's known_hosts parsing and host key verification.

Fixes [#10942](https://github.com/twisted/twisted/issues/10942)
Fixes [#10890](https://github.com/twisted/twisted/issues/10890)
Fixes [#3494](https://github.com/twisted/twisted/issues/3494)
Fixes [#3496](https://github.com/twisted/twisted/issues/3496)
Fixes [#10884](https://github.com/twisted/twisted/issues/10884)
Fixes [#3499](https://github.com/twisted/twisted/issues/3499)

The main thrust of this change is that it adds a new module to conch, `twisted.conch.client.knownhosts`, which provides a structured representation of OpenSSH's `known_hosts` file.  This is a big step in the direction of modularizing conch's support for verifying host keys and storing the results of that verification, although the internal connection APIs are mostly unchanged at this point, they could now easily be adjusted to speak to an abstract interface and still manipulate the user's actual known_hosts entries rather than completely overriding them.

The individual bugs which this change fixes are too numerous to bother describing twice; have a look at the tickets themselves more information.  Suffice it to say that the conch client is now a lot more reliable (although it still has a long way to go).

In addition to making the conch client more stable, the new API allows for some programmatic manipulation of known_hosts files, which might be independently useful.

Additionally, although the review notes some ways in which the tests can still be improved, the tests for this module can be cited as a _much_ better example for conch contributors as to how unit tests within conch should look.  While these tests are covering almost exclusively new code, they are covering functionality which has existed (and been broken) for quite some time.

Author: glyph

Reviewers: exarkun, jml

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
